pkg/externalresource/manager/service.go

```diff
@@ -54,5 +167,199 @@ func (s *Service) CreateResource(
 		return nil, st.Err()
 	}
 
+	// Note that the cache is updated AFTER the metastore is updated.
+	if _, exists := s.cache[request.GetResourceId()]; exists {
+		log.L().Panic("Cache is inconsistent",
+			zap.String("resource-id", request.GetResourceId()))
+	}
+	s.cache[request.GetResourceId()] = resourceRecord
+
+	// TODO: handle the case where resourceRecord.Deleted == true
 	return &pb.CreateResourceResponse{}, nil
 }
+
+func (s *Service) GetPlacementConstraint(
+	ctx context.Context,
+	id resourcemeta.ResourceID,
+) (resourcemeta.ExecutorID, bool, error) {
+	if !s.checkAllLoaded() {
+		return "", false, derror.ErrResourceManagerNotReady.GenWithStackByArgs()
+	}
+
+	logger := log.L().WithFields(zap.String("resource-id", id))
+
+	rType, _, err := resourcemeta.ParseResourcePath(id)
+	if err != nil {
+		return "", false, err
+	}
+
+	if rType != resourcemeta.ResourceTypeLocalFile {
+		logger.Info("Resource does not need a constraint",
+			zap.String("resource-id", id), zap.String("type", string(rType)))
+		return "", false, nil
+	}
+
+	if !s.mu.Lock(ctx) {
+		return "", false, errors.Trace(ctx.Err())
+	}
+	defer s.mu.Unlock()
+
+	record, exists := s.cache[id]
+	if !exists {
+		// Note that although we are not doing cache eviction,
+		// a miss is still a possibility given that we might have
+		// a successful write to metastore but due to network problem
+		// we have treated that write as an error.
+		logger.Info("Resource cache miss")
+		var err error
+
+		startTime := time.Now()
+		record, exists, err = s.accessor.GetResource(ctx, id)
+		getResourceDuration := time.Since(startTime)
+
+		logger.Info("Resource meta fetch completed", zap.Duration("duration", getResourceDuration))
+		if err != nil {
+			return "", false, err
+		}
+		if !exists {
+			return "", false, derror.ErrResourceDoesNotExist.GenWithStackByArgs(id)
+		}
+		s.cache[id] = record
+	} else {
+		logger.Info("Resource cache hit")
+	}
+
+	if record.Deleted {
+		logger.Info("Resource meta is marked as deleted", zap.Any("record", record))
+		return "", false, derror.ErrResourceDoesNotExist.GenWithStackByArgs(id)
+	}
+
+	if !s.executors.HasExecutor(string(record.Executor)) {
+		logger.Info("Resource meta indicates a non-existent executor",
+			zap.String("executor-id", string(record.Executor)))
+		return "", false, derror.ErrResourceDoesNotExist.GenWithStackByArgs(id)
+	}
+
+	return record.Executor, true, nil
+}
+
+func (s *Service) OnExecutorOffline(executorID resourcemeta.ExecutorID) error {
+	select {
+	case s.offlinedExecutors <- executorID:
+		return nil
+	default:
+	}
+	log.L().Warn("Too many offlined executors, dropping event",
+		zap.String("executor-id", string(executorID)))
+	return nil
+}
+
+func (s *Service) StartBackgroundWorker() {
+	ctx, cancel := context.WithCancel(context.Background())
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		<-s.cancelCh
```

Comment:
why don't we let caller directly call cancel? it's concurrent-safe

Reply:
Storing a channel seems more elegant considering we need to handle race with `Stop()`.